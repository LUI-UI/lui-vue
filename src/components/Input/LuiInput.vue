<script lang="ts">
export default {
  name: 'LuiInput',
  inheritAttrs: false
}
</script>
<script setup lang="ts">
import { computed, ref, useAttrs, toRefs, useSlots, h } from 'vue'
import type { PropType } from 'vue'
import type { Rounded, Block, Size, State, StateIcon, Description } from '@/globals/types'
import type { Clear, ModelValue } from './input-types'
import { useInputClasses } from './composables'
import { useGlobalDescriptionClasses } from '../../composables/index'

const props = defineProps({
  size: {
    type: String as PropType<Size>,
    default: 'md'
  },
  rounded: {
    type: [Boolean, String] as PropType<Rounded>,
    default: false
  },
  state: {
    type: [String, Boolean, null] as PropType<State>,
    default: null
  },
  stateIcon: {
    type: [Boolean] as PropType<StateIcon>,
    default: false
  },
  block: {
    type: Boolean as PropType<Block>,
    default: false
  },
  clear: {
    type: Boolean as PropType<Clear>,
    default: false
  },
  prependIcon: {
    type: Boolean as PropType<boolean>,
    default: false
  },
  description: {
    type: [String, null] as PropType<Description>,
    default: null
  },
  modelValue: {
    type: [String, Number] as PropType<ModelValue>,
    default: ''
  }
})

const emit = defineEmits(['update:modelValue'])
const attrs = useAttrs()
const slots = useSlots()
// const LuiInputRef: Ref = ref(null);
const LuiInputRef = ref<HTMLInputElement>()

const {
  inputClasses,
  prependClasses,
  stateIconClasses,
  closeIconClasses,
  wrapperClasses,
  appendClasses
} = useInputClasses(toRefs(props), attrs)
const { descriptionClasses } = useGlobalDescriptionClasses(toRefs(props), attrs)
// const stateIconName = computed(() => {
//   return attrs.disabled
//     ? "forbid-2"
//     : props.state === true
//     ? "checkbox-circle"
//     : props.state === "warning"
//     ? "feedback"
//     : props.state === false
//     ? "error-warning"
//     : "";
// });

// const computedAttrs = computed(() => {
//   const { class: attrClass, style, ...rest } = attrs;
//   return {
//     parent: { attrClass, style },
//     input: rest,
//   };
// });

function clearInput() {
  if (LuiInputRef.value?.value != undefined) {
    LuiInputRef.value.value = ''
  }
  LuiInputRef.value?.focus()
}

const focus = () => LuiInputRef.value?.focus()

defineExpose({
  focus
})

function handleInputEvents(val: any) {
  emit('update:modelValue', val.target.value)
}
const iconSizes = computed(() =>
  props.size === 'xs' ? '12' : props.size === 'sm' ? '16' : props.size === 'xl' ? '24' : '20'
)
// const isDisabled = computed(
//   () => attrs.disabled !== undefined && attrs.disabled === true
// );
type StateIconPathType = {
  warning: string
  feedback: string
  success: string
  disabled: string
}
const stateIconPaths: StateIconPathType = {
  // warning
  warning:
    'M6.455 19L2 22.5V4a1 1 0 0 1 1-1h18a1 1 0 0 1 1 1v14a1 1 0 0 1-1 1H6.455zM4 18.385L5.763 17H20V5H4v13.385zM11 13h2v2h-2v-2zm0-6h2v5h-2V7z',
  // feedback
  feedback:
    'M6.455 19L2 22.5V4a1 1 0 0 1 1-1h18a1 1 0 0 1 1 1v14a1 1 0 0 1-1 1H6.455zM4 18.385L5.763 17H20V5H4v13.385zM11 13h2v2h-2v-2zm0-6h2v5h-2V7z',
  // success
  success:
    'M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm0-2a8 8 0 1 0 0-16 8 8 0 0 0 0 16zm-.997-4L6.76 11.757l1.414-1.414 2.829 2.829 5.656-5.657 1.415 1.414L11.003 16z',
  disabled:
    'M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm0-2a8 8 0 1 0 0-16 8 8 0 0 0 0 16zm4.891-11.477l-8.368 8.368a6.04 6.04 0 0 1-1.414-1.414l8.368-8.368a6.04 6.04 0 0 1 1.414 1.414z'
}

const clearIconTemplate = () =>
  h(
    'svg',
    {
      viewBox: '0 0 24 24',
      fill: 'currentColor',
      width: iconSizes.value,
      height: iconSizes.value
    },
    [
      h('path', { fill: 'none', d: 'M0 0h24v24H0z' }),
      h('path', {
        d: 'M12 10.586l4.95-4.95 1.414 1.414-4.95 4.95 4.95 4.95-1.414 1.414-4.95-4.95-4.95 4.95-1.414-1.414 4.95-4.95-4.95-4.95L7.05 5.636z'
      })
    ]
  )

const stateIconTemplate = (params: any) =>
  h(
    'svg',
    {
      viewBox: '0 0 24 24',
      fill: 'currentColor',
      width: iconSizes.value,
      height: iconSizes.value
    },
    [h('path', { fill: 'none', d: 'M0 0h24v24H0z' }), h('path', { d: params.path })]
  )
</script>
<template>
  <div class="lui-input" :class="wrapperClasses">
    <div class="relative" :class="block ? 'w-full' : ''">
      <input
        ref="LuiInputRef"
        :value="modelValue"
        :class="inputClasses"
        v-bind="$attrs"
        @input="handleInputEvents($event)"
      />
      <span v-if="!!slots.prepend" :class="prependClasses" class="leading-none">
        <slot name="prepend" />
      </span>
      <button v-if="clear && !attrs.disabled" @click="clearInput" :class="closeIconClasses">
        <clearIconTemplate />
      </button>
      <span v-else-if="!!slots.append" :class="appendClasses">
        <slot name="append" />
      </span>
      <span v-else-if="stateIcon && state !== null" :class="stateIconClasses">
        <stateIconTemplate
          :path="
            attrs?.disabled === true
              ? stateIconPaths.disabled
              : state === true
              ? stateIconPaths.success
              : state === false
              ? stateIconPaths.warning
              : stateIconPaths.feedback
          "
        />
      </span>
    </div>
    <p v-if="description !== null" :class="descriptionClasses">
      {{ description }}
    </p>
  </div>
</template>
