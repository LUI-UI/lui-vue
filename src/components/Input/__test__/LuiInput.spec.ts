import { afterEach, beforeEach, describe, expect, it } from 'vitest'
import type { DOMWrapper, VueWrapper } from '@vue/test-utils'
import { mount } from '@vue/test-utils'
import LuiInput from '../LuiInput.vue'

const testValue = 'an example value...'

describe('Default LuiInput', () => {
  let wrapper: VueWrapper<any>
  beforeEach(() => {
    wrapper = mount(LuiInput)
  })
  afterEach(() => {
    wrapper.unmount()
  })
  it('renders properly with default values', () => {
    const expectedDefaultProps = {
      size: 'md',

      rounded: false,

      state: null,

      stateIcon: false,

      block: false,

      clear: false,

      description: null,

      modelValue: undefined,
    }

    const wrapper = mount(LuiInput)

    // Check if the component exists
    expect(wrapper.exists()).toBe(true)

    // Check if the rendered HTML matches the expected HTML
    expect(wrapper.html()).toMatchSnapshot()

    // Check if the default props given correctly
    expect(wrapper.props()).toEqual(expectedDefaultProps)
  })
})
describe('LuiInput class tests', () => {
  let wrapper: VueWrapper<any>
  beforeEach(() => {
    wrapper = mount(LuiInput)
  })
  afterEach(() => {
    wrapper.unmount()
  })
  it('renders correctly when rounded prop is false', async () => {
    await wrapper.setProps({ rounded: false })
    expect(wrapper.classes()).not.contain('rounded')
    expect(wrapper.classes()).not.contain('rounded-lg')
    expect(wrapper.classes()).not.contain('rounded-full')
  })

  it('renders correctly when rounded prop is true', async () => {
    await wrapper.setProps({ rounded: true })
    expect(wrapper.find('input').classes()).contain('rounded-lg')
  })
  it('renders correctly when rounded prop is full', async () => {
    await wrapper.setProps({ rounded: 'full' })
    expect(wrapper.find('input').classes()).contain('rounded-full')
  })

  it('provides full width when block prop is given', async () => {
    await wrapper.setProps({ block: true })
    expect(wrapper.classes()).toContain('w-full')
  })
  it('renders sizes properly', async () => {
    const sizesAndClasses = {
      xs: {
        text: 'xs',
        px: 1.5,
        py: 1,
      },
      sm: {
        text: 'sm',
        px: 2,
        py: 1.5,
      },
      md: {
        text: 'base',
        px: 2.5,
        py: 2,
      },
      lg: {
        text: 'lg',
        px: 2.5,
        py: 2.5,
      },
      xl: {
        text: 'xl',
        px: 3,
        py: 3.5,
      },
    }
    for (const [key, value] of Object.entries(sizesAndClasses)) {
      await wrapper.setProps({ size: key })
      const classes = wrapper.find('[data-testid="lui-input-input"]').classes()
      expect(classes).toContain(`px-${value.px}`)
      expect(classes).toContain(`py-${value.py}`)
      expect(classes).toContain(`text-${value.text}`)
    }
  })
})

describe('LuiInput HTML tests', () => {
  const testClass = 'ri-search-line'
  let wrapper: VueWrapper<any>
  beforeEach(() => {
    wrapper = mount(LuiInput)
  })
  afterEach(() => {
    wrapper.unmount()
  })

  it('renders description properly', async () => {
    await wrapper.setProps({ description: testValue })
    const description = wrapper.find('[data-testid=\'lui-input-description\']')
    expect(description.exists()).toBe(true)
    expect(description.text()).toBe(testValue)
  })

  it('renders prepend slot properly', () => {
    wrapper = mount(LuiInput as any, {
      slots: {
        prepend: `<i class=${testClass}></i>`,
      },
    })
    const prependIcon = wrapper.find(
      '[data-testid=\'lui-input-icon-prepend\'] i',
    )

    expect(prependIcon.exists()).toBe(true)
    expect(prependIcon.classes()).toContain(testClass)
  })
  it('renders append slot properly', () => {
    wrapper = mount(LuiInput as any, {
      slots: {
        append: `<i class=${testClass}></i>`,
      },
    })
    const appendIcon = wrapper.find('[data-testid=\'lui-input-icon-append\'] i')

    expect(appendIcon.exists()).toBe(true)
    expect(appendIcon.classes()).toContain(testClass)
  })
  it('executes clear feature properly', async () => {
    await wrapper.setProps({ clear: true })

    const clearButton = wrapper.find('[data-testid=\'lui-input-clear\']')
    const inputEl = wrapper.find(
      '[data-testid=\'lui-input-input\']',
    ) as DOMWrapper<HTMLInputElement>

    expect(clearButton.exists()).toBe(true)

    inputEl.setValue(testValue)
    expect(inputEl.element.value).toBe(testValue)
    clearButton.trigger('click')
    expect(inputEl.element.value).toBe('')
  })
})

describe('LuiInput state', () => {
  let wrapper: VueWrapper<any>
  beforeEach(() => {
    wrapper = mount(LuiInput, {
      props: {
        'description': testValue,
        'state-icon': true,
      },
    })
  })
  afterEach(() => {
    wrapper.unmount()
  })
  it('applies expected classes and renders proper icon when state prop is true', async () => {
    await wrapper.setProps({ state: true })
    // check the classes
    expect(wrapper.find('[data-testid="lui-input-input"]').classes()).toContain(
      'border-success-500',
    )
    expect(
      wrapper.find('[data-testid="lui-input-description"]').classes(),
    ).toContain('text-success-500')
    // check if proper icon is exists
    expect(wrapper.find('[data-testid="lui-input-state-true"]').exists()).toBe(
      true,
    )
  })
  it('applies expected classes and renders proper icon when state prop is warning', async () => {
    await wrapper.setProps({ state: 'warning' })
    // check the classes
    expect(wrapper.find('[data-testid="lui-input-input"]').classes()).toContain(
      'border-warning-500',
    )
    expect(
      wrapper.find('[data-testid="lui-input-description"]').classes(),
    ).toContain('text-warning-500')
    // check if proper icon is exists
    expect(
      wrapper.find('[data-testid="lui-input-state-warning"]').exists(),
    ).toBe(true)
  })
  it('applies expected classes and renders proper icon when state prop is false', async () => {
    await wrapper.setProps({ state: false })
    // check the classes
    expect(wrapper.find('[data-testid="lui-input-input"]').classes()).toContain(
      'border-danger-500',
    )
    expect(
      wrapper.find('[data-testid="lui-input-description"]').classes(),
    ).toContain('text-danger-500')

    // check if the proper icon is exists
    expect(wrapper.find('[data-testid="lui-input-state-false"]').exists()).toBe(
      true,
    )
  })

  // it("renders icons and border based on the state", () => {
  //   const trueWrapper = mount(LuiInput, {
  //     props: {
  //       description: testValue,
  //       state: true,
  //       "state-icon": true,
  //     },
  //   });
  //   expect(
  //     trueWrapper.find('[data-testid="lui-input-state-true"]').exists()
  //   ).toBe(true);
  //   expect(
  //     trueWrapper.find('[data-testid="lui-input-input"]').classes()
  //   ).toContain("border-success-500");

  //   const warningWrapper = mount(LuiInput, {
  //     props: {
  //       description: testValue,
  //       state: "warning",
  //       "state-icon": true,
  //     },
  //   });
  //   expect(
  //     warningWrapper.find('[data-testid="lui-input-state-warning"]').exists()
  //   ).toBe(true);
  //   expect(
  //     warningWrapper.find('[data-testid="lui-input-input"]').classes()
  //   ).toContain("border-warning-500");

  //   const falseWrapper = mount(LuiInput, {
  //     props: {
  //       description: testValue,
  //       state: false,
  //       "state-icon": true,
  //     },
  //   });
  // expect(
  //   falseWrapper.find('[data-testid="lui-input-state-false"]').exists()
  // ).toBe(true);
  //   expect(
  //     falseWrapper.find('[data-testid="lui-input-input"]').classes()
  //   ).toContain("border-danger-500");
  // });
})

// it("should render correctly based on the rounded prop", () => {
//   // const roundedOptions = {
//   //   false: "",
//   //   true: "lg",
//   //   full: "full",
//   // };
//   const notRoundedWrapper = mount(LuiInput, {
//     props: {
//       rounded: false,
//     },
//   });
//   const roundedWrapper = mount(LuiInput, {
//     props: {
//       rounded: true,
//     },
//   });
//   const roundedFullWrapper = mount(LuiInput, {
//     props: {
//       rounded: "full",
//     },
//   });
//   expect(notRoundedWrapper.classes()).not.contain("rounded");
//   expect(notRoundedWrapper.classes()).not.contain("rounded-lg");
//   expect(notRoundedWrapper.classes()).not.contain("rounded-full");
//   expect(
//     roundedWrapper.find('[data-testid="lui-input-input"]').classes()
//   ).contain("rounded-lg");
//   expect(
//     roundedFullWrapper.find('[data-testid="lui-input-input"]').classes()
//   ).contain("rounded-full");
// });
// it("provides full width when block prop is given", () => {
//   const blockWrapper = mount(LuiInput, {
//     props: {
//       block: true,
//     },
//   });
//   expect(blockWrapper.classes()).toContain("w-full");
//   const wrapper = mount(LuiInput, {
//     props: {
//       block: false,
//     },
//   });
//   expect(wrapper.classes()).not.toContain("w-full");
// });

// it("renders sizes properly", () => {
//   const sizesAndClasses = {
//     xs: {
//       text: "xs",
//       px: 1.5,
//       py: 1,
//     },
//     sm: {
//       text: "sm",
//       px: 2,
//       py: 1.5,
//     },
//     md: {
//       text: "base",
//       px: 2.5,
//       py: 2,
//     },
//     lg: {
//       text: "lg",
//       px: 2.5,
//       py: 2.5,
//     },
//     xl: {
//       text: "xl",
//       px: 3,
//       py: 3.5,
//     },
//   };
//   Object.entries(sizesAndClasses).forEach(([key, value]) => {
//     const wrapper = mount(LuiInput, {
//       props: {
//         size: key,
//       },
//     });
//     const inputEl = wrapper.find('[data-testid="lui-input-input"]');
//     expect(inputEl.classes()).toContain(`px-${value.px}`);
//     expect(inputEl.classes()).toContain(`py-${value.py}`);
//     expect(inputEl.classes()).toContain(`text-${value.text}`);
//   });
// });

// it("renders description properly", () => {
//   const wrapper = mount(LuiInput, {
//     props: {
//       description: testValue,
//     },
//   });
//   const description = wrapper.find("[data-testid='lui-input-description']");
//   expect(description.exists()).toBe(true);
//   expect(description.text()).toBe(testValue);
// });
// it("renders description and border based on the state prop", () => {
//   // const statesAndColors = {
//   //   null: "secondary-600",
//   //   true: "success-500",
//   //   warning: "warning-500",
//   //   false: "danger-500",
//   // };

//   // Check if classes rendered properly for each state
//   const nullWrapper = mount(LuiInput, {
//     props: {
//       description: testValue,
//       state: null,
//     },
//   });
//   expect(
//     nullWrapper.find('[data-testid="lui-input-description"]').classes()
//   ).toContain("text-secondary-600");

//   const trueWrapper = mount(LuiInput, {
//     props: {
//       description: testValue,
//       state: true,
//     },
//   });
//   expect(
//     trueWrapper.find('[data-testid="lui-input-description"]').classes()
//   ).toContain("text-success-500");
//   expect(
//     trueWrapper.find('[data-testid="lui-input-input"]').classes()
//   ).toContain("border-success-500");

//   const warningWrapper = mount(LuiInput, {
//     props: {
//       description: testValue,
//       state: "warning",
//     },
//   });
//   expect(
//     warningWrapper.find('[data-testid="lui-input-description"]').classes()
//   ).toContain("text-warning-500");
//   expect(
//     warningWrapper.find('[data-testid="lui-input-input"]').classes()
//   ).toContain("border-warning-500");

//   const falseWrapper = mount(LuiInput, {
//     props: {
//       description: testValue,
//       state: false,
//     },
//   });
//   expect(
//     falseWrapper.find('[data-testid="lui-input-description"]').classes()
//   ).toContain("text-danger-500");
//   expect(
//     falseWrapper.find('[data-testid="lui-input-input"]').classes()
//   ).toContain("border-danger-500");
// });
// it("renders icons and border based on the state", () => {
//   const trueWrapper = mount(LuiInput, {
//     props: {
//       description: testValue,
//       state: true,
//       "state-icon": true,
//     },
//   });
//   expect(
//     trueWrapper.find('[data-testid="lui-input-state-true"]').exists()
//   ).toBe(true);
//   expect(
//     trueWrapper.find('[data-testid="lui-input-input"]').classes()
//   ).toContain("border-success-500");

//   const warningWrapper = mount(LuiInput, {
//     props: {
//       description: testValue,
//       state: "warning",
//       "state-icon": true,
//     },
//   });
//   expect(
//     warningWrapper.find('[data-testid="lui-input-state-warning"]').exists()
//   ).toBe(true);
//   expect(
//     warningWrapper.find('[data-testid="lui-input-input"]').classes()
//   ).toContain("border-warning-500");

//   const falseWrapper = mount(LuiInput, {
//     props: {
//       description: testValue,
//       state: false,
//       "state-icon": true,
//     },
//   });
//   expect(
//     falseWrapper.find('[data-testid="lui-input-state-false"]').exists()
//   ).toBe(true);
//   expect(
//     falseWrapper.find('[data-testid="lui-input-input"]').classes()
//   ).toContain("border-danger-500");
// });
// it("displays icons properly", () => {
//   const testClass = "ri-search-line";

//   // Tests for prepend slot

//   const prependWrapper = mount(LuiInput, {
//     slots: {
//       prepend: `<i class=${testClass}></i>`,
//     },
//   });
//   expect(
//     prependWrapper.find("[data-testid='lui-input-icon-prepend']").exists()
//   ).toBe(true);
//   expect(
//     prependWrapper
//       .find("[data-testid='lui-input-icon-prepend']")
//       .find("i")
//       .exists()
//   ).toBe(true);
//   expect(
//     prependWrapper
//       .find("[data-testid='lui-input-icon-prepend']")
//       .find("i")
//       .classes()
//   ).toContain(testClass);

//   // Tests for append slot

//   const appendWrapper = mount(LuiInput, {
//     slots: {
//       append: `<i class=${testClass}></i>`,
//     },
//   });
//   expect(
//     appendWrapper.find("[data-testid='lui-input-icon-append']").exists()
//   ).toBe(true);
//   expect(
//     appendWrapper
//       .find("[data-testid='lui-input-icon-append']")
//       .find("i")
//       .exists()
//   ).toBe(true);
//   expect(
//     appendWrapper
//       .find("[data-testid='lui-input-icon-append']")
//       .find("i")
//       .classes()
//   ).toContain(testClass);
// });
