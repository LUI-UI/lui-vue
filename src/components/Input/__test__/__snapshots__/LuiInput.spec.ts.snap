// Vitest Snapshot v1

exports[`Default LuiInput > renders properly with default values 1`] = `
"<div class=\\"lui-input inline-block leading-3 w-48\\">
  <div class=\\"relative\\"><input data-testid=\\"lui-input-input\\" class=\\"peer w-full text-base text-secondary-600 placeholder:text-secondary-400 disabled:placeholder:text-secondary-300 dark:text-secondary-300 dark:placeholder:text-secondary-600 dark:disabled:text-secondary-700 bg-secondary-50 disabled:bg-secondary-100 dark:bg-secondary-900 dark:disabled:secondary-800 outline-none border border-solid border-secondary-200 focus:border-primary-500 disabled:border-secondary-200 focus:ring-4 focus:ring-primary-500/40 disabled:cursor-not-allowed py-2 px-2.5\\">
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
</div>"
`;
